sdks/java/io/csv/src/test/java/org/apache/beam/sdk/io/csv/CsvIOWriteTest.java

```diff
@@ -0,0 +1,534 @@
+        () ->
+            errorPipeline
+                .apply(
+                    Create.of(
+                        Row.withSchema(mapSchema)
+                            .withFieldValue("badfield", ImmutableMap.of("a", "b"))
+                            .withFieldValue("ok", "abcdefg")
+                            .build()))
+                .setRowSchema(mapSchema)
+                .apply(CsvIO.writeRows().to(to)));
+
+    assertThrows(
+        "schema should not contain array field",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(
+                    Create.of(
+                        Row.withSchema(arraySchema)
+                            .withFieldValue("badfield", Arrays.asList("a", "b"))
+                            .withFieldValue("ok", "abcdefg")
+                            .build()))
+                .setRowSchema(arraySchema)
+                .apply(CsvIO.writeRows().withCSVFormat(CSVFormat.DEFAULT).to(to)));
+
+    assertThrows(
+        "schema should not contain row field",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(
+                    Create.of(
+                        Row.withSchema(rowSchema)
+                            .withFieldValue("badfield", rows.get(0))
+                            .withFieldValue("ok", "abcdefg")
+                            .build()))
+                .setRowSchema(rowSchema)
+                .apply(CsvIO.writeRows().to(to)));
+  }
+
+  @Test
+  public void invalidSchemaFields() {
+    String to = tmpFolder.getRoot().getAbsolutePath() + "/somewhere";
+
+    assertThrows(
+        "schema field names should not be empty",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(Create.of(rows))
+                .setRowSchema(ALL_DATA_TYPES_SCHEMA)
+                .apply(
+                    CsvIO.writeRows()
+                        .to(to)
+                        .withSchemaFields(Arrays.asList("", "string", "anInt"))));
+
+    assertThrows(
+        "schema fields should not be an empty list",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(Create.of(rows))
+                .setRowSchema(ALL_DATA_TYPES_SCHEMA)
+                .apply(CsvIO.writeRows().to(to).withSchemaFields(Collections.emptyList())));
+
+    assertThrows(
+        "field names should match schema",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(Create.of(rows))
+                .setRowSchema(ALL_DATA_TYPES_SCHEMA)
+                .apply(
+                    CsvIO.writeRows()
+                        .to(to)
+                        .withSchemaFields(Arrays.asList("doesnotexist", "string", "anInt"))));
+  }
+
+  @Test
+  public void withCompression() {
+    String to = "with_compression";
```

Comment:
nit: It is simpler to reason about the tests if you use a new temporary folder instead of using a special prefix. It also makes file matching easier because then it is just all files in the directory. See https://junit.org/junit4/javadoc/4.12/org/junit/rules/TemporaryFolder.html#newFolder(java.lang.String)